New proposal: JSON Referencing and Identification (JRI)

handrews opened the pull request:
_Note:  This PR will be up for a minimum of 14 days.  If we are nearing that and you want an extra couple of days to comment, please just comment with an estimate of how much additional time you need._

## What this PR is about

As a "new proposal PR", this PR is for _accepting a baseline_ from which to start discussions.  As noted in the [README](https://github.com/json-schema-org/referencing), merging this PR _does not_ indicate an endorsement by the JSON Schema Org, nor does it necessarily mean that it will ever be submitted to the IETF.  Those things depend on the outcome of further discussions of this and other proposals.

Therefore for this and other such PRs (we expect at least one more in this repo):

_Please do!_

* Check that you can get the general idea of the proposal from what's currently written, even if not all details are perfectly clear
* Check for paragraphs/wording that is difficult to understand
* Check for things that seem out of order (I rearranged the order of this proposal in the last few days based on some early feedback to separate the basics, the standalone/interoperable stuff, and the context specification stuff)
* Check for anything that's flat-out wrong
* Make sure I didn't do anything like leave a section titled "Reasons" 🤦 (an early version had this b/c I forgot to come back and make a better title)

_Please do not..._

* Suggest substantive changes.  Please go ahead and file an issue for that!  (if possible, tag it with the `jri` label, but if GitHub won't let you, I'll take care of that)  You do not need to wait for the PR to merge before filing discussion issues
* Raise questions about scope and direction in general.  Please also file issues for those things without waiting for the PR to merge!  As they are larger concerns than any one proposal you should _not_ tag them with this proposal's label

## How to view the file

While there are numerous commits here, that is mostly for the folks who saw an early version of this who want to see how I re-organized it since that version.  Otherwise you can ignore the commits and just review the file as of the most recent commit.

See the [repository README](https://github.com/json-schema-org/referencing/blob/main/README.md) for documentation of the Markdown RFC format.  A few key points:

* The first chunk of the text is a YAML header, so when you look at it as rendered Markdown it looks weird (including some giant headers that don't make sense because they're YAML comments).
* From `--- abstract` onwards, it's Markdown, with some additions
* `{{!RFC1234}}` is an automatic normative reference, `{{?RFC1234}}` is an automatic informative reference, and `{{RFC1234}}` is a reference to something that is explicitly defined in the YAML at the top of the file because the IETF doesn't have automatic entries for whatever it is (e.g. AsyncAPI and OpenAPI)
* `{#foo}` is the equivalent of `"anchor"="foo"` in XML, and `{{foo}}` is the corresponding `xref`
* Footnotes in Markdown become CREFs in the RFC, but that means that if you look at this as rendered Markdown the footnotes show up at the bottom, instead of inline like CREFs do
* You might find it easier to check out my fork and (assuming you have python3 installed) type `make` to get the HTML version built; this will check out a submodule containing the build scripts, install a virtualenv inside that build tree, and run the build for you

handrews committed:
Rearrange sections and fix intro note and refs

No other changes.  Change Log and implementation inventory
are now stored elsewhere, which was why the intro note
needed updating and most of the informative references
were pruned.

handrews committed:
Adjust heading levels, consodidate subsections

handrews committed:
Fix some typos

handrews committed:
Clarify that Notational Conventions defines terms

In addition to the "Definitions" section.

handrews committed:
Move context-specific keyword notes...

...to the section focusing on context specification integration.

handrews committed:
Improve keywords intro, evaluation order

handrews committed:
Attempt a better explanation of locations

and why $defs is part of the spec.

handrews committed:
Reorganize the extending sections

Really, part of it is context-related and part is more fundamental.

handrews committed:
JRI proposal as sent for early feedback.

handrews committed:
More explanation of locations

handrews committed:
Fix submission type

It really should be IETF, "Independent" is something else.

handrews committed:
Reorg the standalone sections to make more sense

handrews committed:
Minor fixes about context-independent documents

handrews committed:
Clarify locators meaning like URLs

handrews committed:
Reorg referencing a bit

Move some things to the standalone and context-sensitive sections.

handrews committed:
Better Web Linking analogies

handrews committed:
Minor tweak to interop section

handrews committed:
Spelling fixes.

handrews committed:
Quote all keywords,

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+Use cases have been compiled from popular tools that support JSON Reference (and sometimes other keywords in this specification) without being full JSON Schema implementations, and from specification requirements regarding non-JSON Schema use.  Notably, some of the tools support additional JRI keywords that were not included in JSON Reference, motivating a larger specification than simply a reviv ...
+
+Discussion of this draft takes place in the [GitHub repository](https://github.com/json-schema-org/referencing) and on the [JSON Schema Slack server](https://json-schema.org/slack).
```
Do you want to specify the slack channel since this link just gets them to the workspace?

handrews replied:
I'll add the channel- it didn't exist when I wrote this and it didn't occur to me to create one.

handrews replied:
Forgot to do this in the previous commit but just pushed a fix now, so I'll mark this resolved.

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+* _data format:_ any specification for structuring data within a document, including but not limited to formal media types
+* _JSON-based format:_ any data format using {{?RFC8259}} JSON as its syntax, including but not limited to formats defined with an `application/json` or `+json` structured suffix media type
+* _JSON Pointer-compatible document:_ any document with a structure that can be unambiguously addressed by JSON Pointer as described in {{jsonpointer}}
+* _root object:_ the object, if any, containing the entirety of a JSON-compatible document[^2]
```
Does this assume the root must always be an object?

handrews replied:
No, there's a CREF/footnote somewhere with open questions about this.

gregsdennis replied:
I'll pull this out into a new issue.

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+* _root object:_ the object, if any, containing the entirety of a JSON-compatible document[^2]
+* _keyword:_ an object property name with specification-defined behavior, as opposed to object properties which are simply data
+* _processing a data format:_ performing any actions required by the format, typically based on the format's structure and keywords, beyond simply parsing the format into an in-memory representation
+* _evaluation path:_ the path, including reference keywords traversed, taken to reach a particular point of processing from the document root at which processing started; the exact nature of this path depends on the reference semantics chosen as defined in {{ref-semantics}}
```
Is this too Schema-specific, or do we expect this to apply for other "context" specifications?

Also, "document root" -> "root object"?  Or maybe change "root object"?  Seems like these are meant to be the same concept.

handrews replied:
There are non-schema JSON Reference tools that use an evaluation path concept, although they don't use that exact name for it.  This is why the JRI spec talks about whether `$ref` appears in the evaluation path (the approach of not showing it is what at least one very popular tool does).

"Keyword" is to distinguish object property names that _do_ something vs those that don't, and yes other specifications besits JSON Schema have keywords in this sense, or are "processed."  Depending on what you're doing with them, OpenAPI and AsyncAPI documents are processed.

Document root and root object are separate because it's not yet clear if documents with a non-object root are something that need to be supported.  If not, then document root and root object will end up being the same thing.

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+For formats that incorporate JRI, "processing" includes resolving relative IRI-references and either associating the resulting IRI with the appropriate resource, or resolving the IRI to the resource with which it is already associated.
+
+[^2]: How should formats without a root object be handled?  Some formats might have a root array.  For interoperability, being able to define how processing starts in terms of a root object is unambiguous.  It is also valid to reference a document that is neither an object or array (early JSON specifications restricted documents to objects or arrays, but that restriction was lifted at least as ea ...
```
It seems to me that `#/1/foo` would be a valid reference in a document that has an array as a root value.

handrews replied:
@gregsdennis I'm not sure what you mean by this.  Yes, that's a valid JSON Pointer for such a document, nothing in this section says that anything is an invalid reference.  The question of requiring root objects or not is not about whether non-object root can be referenced, just a question of whether it's extra complexity and whether that complexity adds anything useful.

gregsdennis replied:
I'm just commenting that arrays at the root are valid, and it seems like they should be considered.  (That's my vote on the issue.)

handrews replied:
Ah, got it, thanks!  They are definitely valid.  I think there might be one or two spots that assume a root object, so we'll need to sort that out, but that's the kind of thing we can talk out in an issue.

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+### Simple and Compound Documents
+
+* _simple document:_ a document that is also a single primary resource
+* _compound document:_ a document consisting of multiple primary resources
```
How often is this seen?  I know that .Net's parser would throw an exception on such documents.  (I disagree with that approach, but it does.)

handrews replied:
All JSON Schemas from 2019-09 onwards with `$id` in a non-document-root object are compound documents.  All OpenAPI 3.1 documents with JSON Schemas that use `$id` are compound documents.  Which means that all bundling use cases that set a `$id` for the bundled documents are compound documents.  I'm not sure exactly how common such things are, but we do know that they exist.

gregsdennis replied:
Ah, okay.  I was considering the case where there were multiple JSON values contained within a single document (e.g. line break delimited), which would definitely cause issues for (at least .Net) parsers.

handrews replied:
Got it.  I think you could use something like [JSON Text Sequences](https://datatracker.ietf.org/doc/html/rfc7464) to construct that sort of multi-document thing, but that would be parsed as JSON Sequence, not as JSON, and the JRI would be used within each JSON document in the JSON Sequence, as there's nowhere outside of that where a keyword could go anyway.

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+* the context specification requires an absolute base IRI defined within each primary resource
+* the context specification requires all IRI-references to be non-relative or fragment-only
+
+Such a base IRI is necessary if the implementation encounters a relative IRI-reference prior to determining a non-relative a base IRI defined within the content.  It is assumed that any provided base IRI has been determined in accordance with the relevant specifications.
```
"prior to determining a non-relative ~a~ base IRI"  (extra "a")

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+#### Exclusion of "$comment"
+
+For compatibility with JSON Schema {{?I-D.bhutton-json-schema-01, Section 8.3}}, JRI requires that a `"$comment"` keyword present in the same object as any JRI keywords MUST NOT impact the semantics of those JRI keywords.  In particular, `"$comment"` MUST NOT be used as a semantic disambiguator as described in {{disambiguator}}.
+
```
Do we need the double quotes _and_ the formatting around "$comment" or does the formatting alone suffice?  (Similar for other places in the doc.)

handrews replied:
I had it as formatting-only until the very last minute, but there are two problems with that:

* Minor problem: the formatting is not really dramatic if you're just kind of skimming things
* Major problem: the formatting does not show up at all in the .txt, .pdf, and probably not the "htmlized" format ("htmlized" is the one with the red bar across the top)

However, I do really prefer to have the formatting on the nice HTML output, and it doesn't cause problems on any other format.  So the quotes are actually the mandatory side, and the formatting is a nice-to-have in the one format where it works.

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+The value of the `"$id"` property MUST be a string, and MUST be a valid IRI-reference as defined by {{!RFC3987, Section 2.2}}, and MUST NOT contain a fragment.
+
+The object containing the `"$id"` property MUST be considered to be a primary resource and to be identified by the IRI produced by resolving the IRI-reference against the current base IRI.  This IRI MUST be considered the base IRI for the newly identified primary resource, in accordance with {{!RFC3986, Section 5.1.1}}.
```
"MUST be considered to be a primary resource and to be identified" is a bit muddy.  Maybe repetition here would be good.  E.g.

> ... MUST be considered to be a primary resource and MUST be identified by the IRI produced...

handrews replied:
How about "MUST be considered to be a primary resource that is identified by the IRI produced..."?

DavidBiesack replied:
or
> The object containing the `"$id"` property MUST be considered to be a primary resource. The object containing the `"$id"` property MUST be identified by the IRI produced by resolving the IRI-reference against the current base IRI.

gregsdennis replied:
I don't read what you put ☝️ and the original text as the same thing, but if that's what you meant, it does read better.

handrews replied:
What I'm going for is that the IRI produced by `$id` MUST be considered an IRI for the primary resource in question.  What I did _not_ mean to imply is that _only_ that IRI can be considered a valid identifier for it.  Does that help, or just muddy it further?

gregsdennis replied:
or

> The object containing the `"$id"` property MUST be considered to be a primary resource **identifiable** by the IRI produced...

handrews replied:
@DavidBiesack 

> The object containing the "$id" property MUST be identified by the IRI produced by resolving the IRI-reference against the current base IRI.

This suggests to me that it can _only_ be identified by that IRI, which is not the case.  There might be valid IRIs to use that come from outside of the context of the document- it's not really within the scope of this spec but it doesn't seem like we should imply that all other possible IRIs are unusable.  Does that make sense?  Please take a look at the updated wording and see if it gives a better/more clear impression.

gregsdennis replied:
(There's also language elsewhere that states a single resource may be identifiable by multiple IRIs, but a single IRI can identify at most a single resource.)

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+#### `"$defs"`
+
+The value of the `"$defs"` property MUST be an object, which MUST have objects as the values of all of its properties.  JRI keywords MUST be allowed in these objects.  Aside from their use in JSON Pointers or similar structural identifiers, the property names within the `"$defs"` object MUST NOT be considered to impose specific semantics to the property values.
```
What if I wanted to do

```json
{
  "$defs": {
    "anything-goes": true
  },
  ...
}
```

JSON Schema would currently accept this (as unhelpful as it might be), but the value of `/$defs/anything-goes` isn't an object.

handrews replied:
Good catch!  I'll have to think on this but agree this should be fixed before merging.  Boolean schemas don't really translate.

There is some language (I think, unless I took it out) about the value of `$defs` being the same as the root object, which _would_ encompass boolean schemas if we made it something like the "root entity" (which we'd have to do if we're allowing arrays anyway... except maybe the root array isn't really the thing we'd want? meh, I will think about it).

There is something here about the values of the `$defs` object being referenceable things.  The trick is how to say that in a meaningful yet flexible way.  🤔

gregsdennis replied:
What if you do what you've done in other places: JRI says it can be anything, but context specs can further restrict it?  That would cover a hypothetical system that wanted to use `$ref` and `$defs` to break up plain JSON but still allow JSON Schema to restrict the values to objects and booleans.

handrews replied:
Yeah, it might be that simple.  I'll try to write it up and see if it works.

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+Context specifications that use keyword-level semantics SHOULD ensure that, in the absence of keywords adjacent to JRI reference keywords, object-level JRI reference semantics can be assumed without changing the semantics of the document according to the context specification.  This requires that the difference in evaluation paths not impact the context semantics.
+
+For example, in JSON Schema, replacing an object that contains only `"$ref"` is guaranteed to not change the validation outcome, and only changes the evaluation path of annotations.  Since the presence of `"$ref"` in the evaluation path is used to determine whether keywords are present adjacent to `"$ref"`, omitting it when no such adjacent keywords are present is safe.  While JSON Schema specifi ...
```
> Since the presence of `"$ref"` in the evaluation path is used to determine whether keywords are present adjacent to `"$ref"`, omitting it when no such adjacent keywords are present is safe.

This is not the reason `$ref` is included in the evaluation path.  The purpose is to show that a reference was followed and resolved.  Omitting `$ref` from the evaluation path is not an option.

Granted, the semantics of _performing_ the evaluation operates in this way, but reporting it in the output does not.

handrews replied:
@gregsdennis this paragraph was motivated by tooling other than JSON Schema, so the JSON Schema output process is not relevant.  I can clarify that if you think it is necessary.

handrews replied:
I don't want to bring up JSON Schema-specific things unless absolutely necessary, though.  The document is long as it is and JSON Schema-specific things are covered in JSON Schema.

Nothing in this document changes how anything works in JSON Schema.

handrews replied:
@gregsdennis OK I've spent a good bit more time staring at the paragraph and trying to figure out what to do with it.  I decided to get rid of that JSON Schema example, and instead address the concern more generally with a subsection in the context-specific part of the spec.  Please see if you think this is an improvement (I'm really not sure if the new language is clear enough, or if it is at least enough to move further discussion into issues).

gregsdennis replied:
The text in the new section looks good.

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+### Ensuring consistent reference semantics
+
+A context-independent format MUST NOT allow any non-JRI-reference keywords in the same object as any JRI reference keyword.[^11]
```
Does this mean that `foo` isn't allowed to be a sibling of `$ref`?  How will that work in JSON Schema where we now allow sibling keywords?

gregsdennis replied:
I see [line 397](https://github.com/json-schema-org/referencing/pull/2/files#diff-a52f2848d61740c55c6f7756031d478fc004b2e4a369b6456e5eec2a1563988dR397) but the wording here seems to contradict that.

handrews replied:
@gregsdennis JSON Schema is not a context-independent format.

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+A common family of use cases involves reading one or more JRI-using documents and transforming them in a way that makes them usable by a wider range of tooling.
+
+#### Reference removal
```
Should there be mention of recursive/circular references in this section?

handrews replied:
It's under Security Considerations.

handrews replied:
(but maybe it should get an xref here now that you mention it)

DavidBiesack replied:
@handrews add an xref here

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+#### Bundling with stable references
+
+A set of documents that each set `"$id"` in their root objects, and that only refer to each other using the primary resource identifiers set that way (plus any appropriate fragment syntax) can be bundled into a single compound document that satisfies all references internally without the need to edit them.
```
A set of documents that each ~set~ define `"$id"`

Removes a little redundancy and the multi-definition usage of "set"

DavidBiesack replied:
"set that way"  confuses me. what way? Do you mean primary resource identifiers that correspond to the `"$id"` values?

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+Any document that can be converted as described in the previous paragraph could equivalently evaluate syntactically valid JSON Pointers by substituting its own object and array definitions for JSON Objects and Arrays, and adjusting the string comparison rules as needed if it does not use Unicode.
+
+Note that a many documents in a format can be JSON Pointer-compatible, even if the format in general is not.  For example, YAML is not compatible due to allowing non-string property values, but the very commonly used JSON-compatible subset of YAML is, of course, JSON Pointer-compatible.  The proposed YAML media type registration {{?I-D.ietf-httpapi-yaml-mediatypes-03, Section 3.3}} addresses this ...
```
> Note that a many documents in a format

Grammar?

handrews replied:
🤦

handrews committed:
Various review feedback bits

The most significant change here is reworking the discussion of
falling back to object-level semantics.

handrews commented:
@gregsdennis I think I have addressed each of  your comments with either a change, an explanation, or a question as to how to proceed.  Please let me know if I missed anything, and thank you for the speedy review!

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+While {{?RFC8288}} web linking solves this problem in a more general way, a simplified, concise, and self-contained linking and identification system is better-suited to many specifications.  This allows those specifications to avoid complexity, and to avoid constraining their syntax to match any one of the many competing web linking-enabled JSON formats.
+
+To support a variety of use cases, JRI supports trade-offs between generic interoperability and tight integration with the context in which it is used.  JRI defines an interoperable subset of its features for standalone use, while encouraging specifications that need reference and/or identification features to incorporate as much or as little of JRI as suits their needs.  JRI also defines mechani ...
```
some caution here: will "incorporate as much or as little of JRI"  opens a door to à la carte standards compliance; i.e. use some of JRI, but deviate from it elsewhere.

gregsdennis replied:
Maybe having a statement like, "A context specification MUST NOT include behaviors that conflict with JRI behaviors even if the context specification elects not to utilize those JRI behaviors."  This would prevent a context spec from including `$ref` but defining its own `$defs`.

handrews replied:
This is addressed in [§4.4 Incorporating a Subset of JRI](https://github.com/json-schema-org/referencing/pull/2/files/8bc3160e0eb19907a85d5bba3ce0ae401ba5f87c#diff-a52f2848d61740c55c6f7756031d478fc004b2e4a369b6456e5eec2a1563988dR407), which was originally much earlier in the document.

I'll have to figure out where the best home for that is.  I don't want to overwhelm with too much context specification-related information before defining the basics, but this should probably be mentioned early on.  Perhaps even tacking on a "as long as they do not redefine the unused portion" or something similar in this paragraph (the intro paragraph shouldn't have any RFC 2119 keywords, IMHO, as it's just a casual overview).

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+The terms "primary resource" and "secondary resource" in this specification are to be interpreted as in {{!RFC3986, Section 3.5}}.
+
+Due to the distinction that this specification makes between documents and resources, the term "same-resource reference" is to be interpreted in the same way as "same-document reference" as defined in {{!RFC3986, Section 4.4}}.  The term "same-document reference" is not used in this specification.
```
since "same-resource reference" is a new term and you are not using "same-document reference" in this RFC, should "same-resource reference" be moved to Definitions below?

handrews replied:
@DavidBiesack I'm a little unclear on when something goes under "Notation conventions" and when it goes under "Definitions" - I certainly have no objection to moving it, I just have no idea which is correct.  Terms that aren't redefined definitely go in "Notation conventions" and entirely new terms definitely go in "Definitions".  Maybe I should just mention it in both?

DavidBiesack replied:
since the notations section is so small, mentioning in both seems redundant. I prefer to see _new_ terms defined in Definitions, use of existing concepts/conventions in the prior, and this feels like a term/definition

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+#### The "$" prefix
+
+JRI keywords MUST start with a "$" character.  Context specifications MAY define their own "$"-prefixed keywords, however this is NOT RECOMMENDED as future specifications that extend JRI as discussed in {{extending}} will use that prefix.[^8]
```
Can `$$` (or `$-`) be reserved for context specifications?

handrews replied:
@DavidBiesack Can you open an issue for this?  And if possible give some use cases for it?  I'm not against this, but I think it warrants a more visible discussion of its own than it will get in this PR.

DavidBiesack replied:
Thanks - see https://github.com/json-schema-org/referencing/issues/6

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+For compatibility with JSON Schema {{?I-D.bhutton-json-schema-01, Section 8.3}}, JRI requires that a `"$comment"` keyword present in the same object as any JRI keywords MUST NOT impact the semantics of those JRI keywords.  In particular, `"$comment"` MUST NOT be used as a semantic disambiguator as described in {{disambiguator}}.
+
+However, `"$comment"` is not itself considered to be a JRI keyword.  JRI does not place any requirements on the usage, data type, syntax, or other semantics of `"$comment"` in context specifications.  JRI implementations MUST NOT require that a `"$comment"` adjacent to JRI keywords conform to JSON Schema's `"$comment"` keyword, unless explicitly required to do so by a context specification.
```
I see this a possibly problematic -- if one context specification suddenly decides `"$comment"` must confirm to JSON Schema's `"$comment"` that could limit interoperability with other context specifications which have different treatment. Is it too limiting to simply adopt JSON Schema's `"$comment"` keyword semantics a sa MUST here?

handrews replied:
@DavidBiesack JSON Schema's `$comment` keyword is bizarrely controversial.  I don't understand it at all, but some people really hate it.  I'm kind of trying to make sure it doesn't get stepped on without causing the anti-`$comment` brigade to come after me.

I would have no objection to including it, but I know there would be objections.  It might be something that has to get sorted out in an issue as I doubt we could sort it out within this PR in a timely manner.

handrews replied:
See #9.  And to some degree, #10.

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+1. Locations (`"$defs"`)
+1. References (`"$ref"`, `"$extRef"`)
+
+The logic of this ordering is that primary resources must be identified first as all other keywords depend having the correct base IRI.  Next, secondary resources can be identified once the primary resource is known.  Finally, nested locations which can be reference targets or have additional identifier keywords can be discovered within the primary resource.
```
maybe  "resolved" instead of "discovered"?
Is "discovered" the same as "newly identified" in the `"$id"` section below? (if so, use one term consistently)

handrews replied:
@DavidBiesack what I'm going for is that "discovery" means finding identifiers, while "resolution" means either relative IRI-reference resolution (but only if specifically noted), or more often, resolution of a reference (`$ref`/`$extRef`).  I didn't particularly mean for "discovery" (much less "newly identified") to be a formal term.  In JSON Schema we talk about "loading a schema" and finding `$id` and `$anchor` during that process.   (I'm not quite sure what my point here is or what to do about this, I'll think on it more).

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+### Location
+
+Location behavior is necessary when not all locations within a data format are valid reference targets, or when object properties that appear to be JRI keywords are only actually JRI keywords in certain locations.  In formats where most locations cause automatic actions (such as applying a schema to an instance), location behavior can also create areas where the automatic actions do not apply.  T ...
```
This para is unclear because "Location behavior" is not yet defined. Perhaps start with the problem ("not all locations within a data format are valid reference targets") then introduce "location behavior" as a solution, preferably with an example. Similarly you jump right in and use define "automatic actions" (with an example) but then not used anywhere else, so that also detracts from the goal of the para.

handrews replied:
@DavidBiesack thanks, these sections used to be in a different order and I missed this, will fix and ping you once I have!  I appreciate the careful reading.

DavidBiesack replied:
I still think this section is not clear enough to someone coming to this draft without background.
Define "Location behavior" above (new term?), or perhaps use "location-specific behavior" which I think is clearer.

In some data formats, not all possible locations are valid reference targets, or object properties that appear to be JRI keywords are only actually JRI keywords in certain locations. In such cases, _location-specific behavior_ is necessary. In formats where most locations cause automatic actions (such as applying a schema to an instance), location behavior can also define areas where the automatic actions do not apply. ...

(I suggest using "define" instead of "create" - the format does not "create" locations in a document.

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+### Location
+
+Location behavior is necessary when not all locations within a data format are valid reference targets, or when object properties that appear to be JRI keywords are only actually JRI keywords in certain locations.  In formats where most locations cause automatic actions (such as applying a schema to an instance), location behavior can also create areas where the automatic actions do not apply.  T ...
```
> when object properties that appear to be JRI keywords are only actually JRI keywords in certain locations

this also confused me and appears at first to be a contradiction. Perhaps rephrase it such that the use of a keyword may map to a JRI keyword in some location, but not map to a JRI keyword in others (or something to that effect -- that's what I get out of "in certain locations")

gregsdennis replied:
I think this addresses the appearance of JRI keywords in places where they're not expected to behave like keywords.  For example, in the JSON Schema Core meta-schema

```json
{
  ...,
  "properties": {
    ...,
    "$ref": { ... },
    ...
  },
  ...
}
```

Here, `$ref` is merely a property name to be found in an instance.  It's not expected to have JRI semantics.

handrews replied:
@DavidBiesack it's about what @gregsdennis mentioned.  This was a problem with a literal reading of JSON Schema draft-04 plus JSON Reference draft-03, which was that you could't even use `$ref` in a place that _was not a keyword_.  This is part of why I defined "keyword" as distinct from just plain "object property name".

I'm open to suggestion for better wording, and will also think on this as I do the next set of revisions for this PR even if there are no further suggestions.  Clearly it needs some improvement!

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+The passive nature of location behavior makes it particularly suitable to be defined directly by context specifications, as noted in {{context-loc}}.  However, not all uses of JRI involve a context specification.
+
+JRI reserves the `"$defs"` keyword in order to make no-automatic-action location behavior available within the `"$"`-prefixed namespace.  This keyword also enables safe interoperable bundling by standalone tools as described in {{bundling-interop}}.[^70]
```
Is
> within the `"$"`-prefixed namespace.

the same thing as "for JRI keywords" ? ("namespace" is not defined in this proposal, or even used anywhere else, so it raises a question about what it means, such as "How does this namespace differ from a vocabulary?"

handrews replied:
Ah, yes, this was a last-minute change (can you tell I rewrote all the location stuff very late in the game? 😅 ).  I'll take a look at this and the general syntax stuff that introduces the `$` prefix and make sure it's all consistent, thanks!

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+#### `"$defs"`
+
+The value of the `"$defs"` property MUST be an object, which MUST have objects as the values of all of its properties.  JRI keywords MUST be allowed in these objects.  Aside from their use in JSON Pointers or similar structural identifiers, the property names within the `"$defs"` object MUST NOT be considered to impose specific semantics to the property values.
```
The MUST also implies a `"$defs"` object may not have a `"$comment": "string"`  -- is that intentional?

handrews replied:
@DavidBiesack yes, that's intentional, and related to the whole bit about things that look like JRI keywords appearing in places that aren't keywords at all.  You can put `$comment` next to `$defs`, but inside `$defs` it's just a name of a definition.  I'd call it bad form to name your definition after a keyword but people do weird things.  A context specification can forbid that if they want.

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+#### `"$extRef"`
+
+To avoid overloading the behavior of `"$ref"` or relying on non-interoperable context-specific behavior, `"$extRef"` (for "extensible reference") allows specifying additional metadata as well as alternative mechanisms for secondary resource identification.[^90]
```
This proposal does not define the format of `"$extRef"` -- is it a string? Could it be an object? Is that the intent? For example, if the resource at the named URI supports multiple media types, can I define an `"$extRef"` and also pass additional HTTP headers such as `"Accept: application/json"` so we don't get back `Content-Type: text/html"` instead if the URI does not have an implicit type via `.json` in the path (though of course that URI is not a guarantee of the representation format).
Could the `$extRef` contain other explicit behaviors, such as JSON transformations?

handrews replied:
@DavidBiesack 🤦  It's supposed to be an object, apparently I spliced out that paragraph, thanks for catching it.

The HTTP headers approach is one possibility.  `$extRef` is intentionally under-specified because I'm not entirely sure how it should work and know that there are some ideas floating around out there.  Assuming there's interest in building on this submission once we're able to evaluate other submissions that come in, I suspect we'll spend substantial time on fleshining out `$extRef`.

But I will fix the part about defining it as an object.

I'm generally vehemently against doing anything based on file extensions, but that's more of a personal preference than a considered recommendation, and also the sort of thing can be debated in issues once we have a baseline proposal merged.

DavidBiesack replied:
I agree - we can't rely on extension - that's why we may need to pass `Accept:`.

DavidBiesack replied:
@handrews the update still does not specify that `"$extRef"` is an object. Add:

```
The value of the `"$extRef"` property MUST be an object.
```
Are there other constraints? i.e. MUST it contain a `"$ref"` string (or perhaps another nested `"$extRef"` ?)

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+##### Object-level semantics {#object-level}
+
+With object-level semantics, an object containing a JRI reference keyword is effectively replaced with its target.  This results in an evaluation path in which the reference keyword does not appear.
```
Perhaps you get to this later, but reference cycles should be addressed. The in-line replacement is not recursive here, so while the reference keyword is replaced with the referenced object, the replacement does not recurse into the new object.  I.e. a schema that contains a property that is also defined with the same schema does not result in blowing up.

handrews replied:
They are addressed later.  I thought I'd added an xref at the first place it was relevant but apparently not, I'll take a look again.

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+##### Object-level semantics {#object-level}
+
+With object-level semantics, an object containing a JRI reference keyword is effectively replaced with its target.  This results in an evaluation path in which the reference keyword does not appear.
```
> an object containing a JRI reference keyword is effectively replaced with its target.

I think this is not correct. In a JSON Schema

```json
        {
            "title": "Feature A",
            "properties": {
                "enabled": {
                    "$ref": "#/$defs/enabledToggle",
                    "default": true
                }
            }
        },
```

the proposal implies the entire value of the `enabled` object  (the "object containing a JRI reference keyword") is replaced with the object at the `$ref`, which would drop the` "default": true` value.

Same is true for [OAS 3.1 Reference Objects](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#referenceObject).

In the clause "an object containing a JRI reference keyword", does _object-level semantics_ mean "an object containing __only__ a JRI reference keyword", and hence the schema example I gave above is a keyword-level semantics? The two terms are not well defined, so it would help to define them, how they differ, and how an observer knows which one applies in a arbitrary context.

handrews replied:
JSON Schema uses keyword-level semantics.  As does the OAS 3.1 Reference Object (although I had not actually remembered that until someone pointed it out the other day).

DavidBiesack replied:
thanks -- thus I think:
> The two terms are not well defined, so it would help to define them, how they differ

jdesrosiers replied:
@handrews 

> JSON Schema uses keyword-level semantics. As does the OAS 3.1 Reference Object 

I wouldn't consider OAS 3.1 as keyword semantics. Keyword semantics imply that the resolved reference would replace the keyword value, but in OAS, a Reference Object merges the referred value with the additional keywords and replaces the Reference Object with the resulting value. That makes it a variation of a object semantics rather than keyword semantics.

(This is how I understood the keyword vs object semantics concept, but you defined them so please tell me if I understood wrong)

handrews replied:
@jdesrosiers yeah that might be more accurate.  I haven't thought about this much as I'd forgotten about the `summary` and whatever else it is until you brought it up on the setup PR for this repo.  Thanks for pointing this out.

DavidBiesack replied:
@handrews please define _object-level semantics_ and _keyword-level semantics_ in the **Definitions** section

handrews committed:
Mention specific slack channel

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+The property names in the value of `"$defs"` are known to be non-keyword properties as this object syntax is defined by JRI.  These property names therefore MUST be considered exempt from the other requirements in this section regarding property names.
+
+A context-independent format MUST NOT allow object properties with names matching JRI keywords and values that appear valid for that keyword that are not intended to function as JRI keywords anywhere within the format.
```
Is
> and values that appear valid for that keyword

necessary here? By adding that restrictive clause, the proposal implies that :

> A context-independent format MAY ALLOW object properties with names matching JRI keywords and values that DO NOT appear valid for that keyword

compare to just

> A context-independent format MUST NOT allow object properties with names matching JRI keywords that are not intended to function as JRI keywords anywhere within the format.

The next sentence leads me to think you need the clause and the implied converse is true  - I just want to be sure it's intentional. It may help to define "appear valid" precisely; the next para implies it is only by data type and no other validation (for example in the context of JSON Schema, there is no checking agains an `enum` or other constraint). This seems very loose definition of "appear valid" feels problematic.

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+### Finding JRI identifiers
+
+A context-independent format MUST NOT allow JRI identifiers anywhere other than its root object or under a `"$defs"` keyword appearing (recursively) in the root object.[^10]
```
(for my understanding - something like OpenAPI is a context-_dependent_ format, so this does not apply there, correct?)

handrews replied:
Correct.  OpenAPI defines exactly where `$ref` can appear in an OAS document, and what the semantics of those appearances are.

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+* Be able to associate a document, including non-JRI documents that could be JRI reference targets, with an IRI
+* Locate all JRI identifiers (primary and secondary resources) within any document and automatically include them in the cache
+* Locate all JRI references and resolve any relative IRI-references to full IRIs in the cached representation
+* In its default configuration, treat cache misses as errors, since JRI reference IRIs are be assumed to be locators
```
`IRIs are be assumed` -> `IRIs are assumed` (remove "be")

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+### The "jri-keywords" media type parameter
+
+The value of the `jri-keywords` parameter MUST be a case-sensitive comma-separated list of JRI keyword names, which MUST NOT contain whitespace.  These keywords MAY be in any order.  This indicates that only the listed keywords are used in the document. In the absence of this parameter, an implementation MUST assume that any JRI keyword can be used.
```
the items SHOULD be unique (i.e. no duplicates)?
or define this as a "set" instead of a "list"?

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+An implementation MUST guard against infinite reference loops.
+
+Processing rules for context specifications MAY make it safe to have cyclic references by creating conditions under which a reference will or will not be followed.
```
Why MAY instead of MUST?

gregsdennis replied:
JSON Schema allows cyclic references (e.g. a schema for a linked list or binary tree), but also requires the guard mentioned above.  Not all cyclic references are necessarily infinite.

handrews replied:
@gregsdennis I think @DavidBiesack is asking why, if cyclic references are allowed by a context specification, we don't require the context specification to define such safety rules.

This is a MAY because I'm not sure a MUST is feasible.  @DavidBiesack if you want to explore this further I think it would make a good issue to file.  I haven't really worked through the possibilities and that feels to big for the initial acceptance PR.

DavidBiesack replied:
I submitted issue #8

DavidBiesack commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,536 @@
+
+Processing rules for context specifications MAY make it safe to have cyclic references by creating conditions under which a reference will or will not be followed.
+
+Cyclic references cannot safely be processed in a way that is both interoperable and context-independent.  Seeing a specific reference while processing the target of that same reference MUST be considered an error by context-unaware implementations.
```
can you clarify this, especially the part about  "Seeing a specific reference" -- does that mean before or  after resolving the base URI, etc.? 
For example in our OAS documents, we define a schema for an [RFC 7807](https://www.rfc-editor.org/rfc/rfc7807#section-3.1) problem object (called `apiProblem`) and extended it with a `problems` array of `apiProblem` objects, in this case defined with an array whose `items` are`$ref: '#/components/schemas/apiProblem'`  Does that violate this clause and MUST be considered an error?
We also extract (debundle) these schemas into context-independent files.

handrews replied:
I think this was a casualty of me reorganizing sections.  It's an error if you are trying to remove references, but in other situations it's completely fine as you note.  I'll rework this.

DavidBiesack reviewed (commented):
Thanks for publishing this and soliciting feedback. I hope my comments are helpful.

handrews committed:
Make clear that partial adoption stays compatible

You can't adopt "$ref" but redefined "$id" to do something else.

handrews committed:
Move same-resource reference def to definitions

Since it is not a direct adoption of terminology from another spec.

handrews committed:
Be more consistent about identifier "discovery"

handrews committed:
Keywords in param SHOULD be unique; wording fixes

handrews committed:
Clarify required cyclic error detection conditions.

handrews commented:
@DavidBiesack I _think_ I have addressed all of your comments that we did not move off to separate issues.  Please take a look.  Since this has been open for 14 days, once you have signed off I will merge this unless something else has come up.  We will continue to discuss things in issues and all are welcome to file such issues after the merge.

handrews committed:
Wording improvement for "$id" behavior.

handrews committed:
Relax syntax of "$defs"

It needs to be able to accomodate boolean schemas, and formats
in which any value can be a reference target.

handrews commented:
OK, @gregsdennis I _think_ I have now addressed everything- thanks for the IM about the ones that GitHub had "conveniently" hidden.

handrews commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,541 @@
+
+### Document set transformations
+
+A common family of use cases involves reading one or more JRI-using documents and transforming them in a way that makes them usable by a wider range of tooling.  Note that circular references can cause problems for these use cases, as discussed in {{infinite}}.
```
@DavidBiesack I think you're looking at an outdated diff.  Here is the xref that you requested.

handrews committed:
$extRef is an object, with contents TBD

handrews committed:
Define object-level & keyword-level semantics

handrews committed:
Clarify location/location-definng behavior

handrews commented:
@DavidBiesack thanks for the speedy review!  I think I have addressed everything you raised- I have no idea how I missed the `$extRef` comment in particular, this PR has gotten rather confusing to follow in the UI (I missed a comment or two of @gregsdennis's as well despite going over everything multiple times).  Please let me know if this is insufficient or if you notice other things I have missed.  I did kind of rush this most recent update.

gregsdennis commented on draft-handrews-jri.md:
```diff
@@ -0,0 +1,549 @@
+
+The value of the `"$id"` property MUST be a string, and MUST be a valid IRI-reference as defined by {{!RFC3987, Section 2.2}}, and MUST NOT contain a fragment.
+
+The object containing the `"$id"` property MUST be considered to be a primary identifiable by the IRI produced by resolving the IRI-reference against the current base IRI.  This IRI MUST be considered the base IRI for the primary resource that it identifies, in accordance with {{!RFC3986, Section 5.1.1}}.
```
```suggestion
The object containing the `"$id"` property MUST be considered to be a primary resource identifiable by the IRI produced by resolving the IRI-reference against the current base IRI.  This IRI MUST be considered the base IRI for the primary resource that it identifies, in accordance with {{!RFC3986, Section 5.1.1}}.
```

gregsdennis reviewed (approved):
I have one little nit remaining (omitted word), but overall I like this.

DavidBiesack reviewed (approved):
Thanks @handrews - I have no remaining changes

handrews committed:
Fix missing word.

handrews commented:
@DavidBiesack @gregsdennis thanks!  I have fixed the word 🤦
